src/block-cursor.ts

```diff
@@ -212,11 +232,50 @@ function measureCursor(cm: CodeMirror, view: EditorView, cursor: SelectionRange,
       // include the second half of a surrogate pair in cursor
       letter += view.state.sliceDoc(head + 1, head + 2);
     }
+
+    // Calculate actual character width by measuring the rendered text
+    let charWidth = 8; // default fallback
+
+    // Special handling for newlines and end-of-line
+    let actualLetter = view.state.sliceDoc(head, head + 1);
+    if (!actualLetter || actualLetter == "\n" || actualLetter == "\r" || head >= view.state.doc.length) {
+      // Newline or end of document: use narrow cursor
+      const fontSize = parseInt(style.fontSize) || 16;
+      charWidth = fontSize * 0.15; // Very narrow for newlines
+    } else {
+      // Try to measure from the original DOM node before traversal
+      if (originalDomAtPos && originalDomAtPos.node instanceof Text) {
+        const range = document.createRange();
+        const textNode = originalDomAtPos.node;
+        const offset = originalDomAtPos.offset;
+
+        if (offset < textNode.length) {
+          try {
+            range.setStart(textNode, offset);
+            range.setEnd(textNode, Math.min(offset + 1, textNode.length));
+            const rect = range.getBoundingClientRect();
+            if (rect.width > 0 && rect.width < 100) {
+              charWidth = rect.width;
+            }
+          } catch (e) {
+            // Range measurement failed, will use fallback
+          }
+        }
+      }
+
+      // Fallback: use font-based estimation
+      if (charWidth <= 0 || charWidth >= 100) {
+        const fontSize = parseInt(style.fontSize) || 16;
+        charWidth = fontSize * 0.6; // reasonable default for most characters
+      }
+    }
+
     let h = (pos.bottom - pos.top);
     return new Piece((left - base.left)/view.scaleX, (pos.top - base.top + h * (1 - hCoeff))/view.scaleY, h * hCoeff/view.scaleY,
+                     charWidth/view.scaleX,
                      style.fontFamily, style.fontSize, style.fontWeight, style.color,
                      primary ? "cm-fat-cursor cm-cursor-primary" : "cm-fat-cursor cm-cursor-secondary",
-                     letter, hCoeff != 1)
+                     letter, true) // Always use transparent letter to preserve RTL character connections
```

Comment:
To preserve intra-word "letter connection" in connected scripts like Arabic (this fix has nothing to do with RTL per se so this is an inaccurate comment)